README.md
# pxh-chrome 3.0.3

Application chrome for the Predix UI AppHub

## [Live Demo](https://github.build.ge.com/pages/hubs/pxh-chrome-demo)

pxh-chrome is the CSS/HTML layer of [ui-app-hub](https://github.build.ge.com/hubs/ui-app-hub), a front-end architecture that allows you to build and deliver separate [microapps](https://github.build.ge.com/hubs/ui-microapp) to your users through a shared navigation portal. Adhering to a mobile-first mindset, pxh-chrome offers the following components:

* `pxh-drawer`, a fixed lefthand drawer that contains application navigation, user profile information, and login controls
* `pxh-drawer-toggle` - a "hamburger" that allows a user to toggle the navigation drawer open or closed
* `pxh-navigation`, a linked list that works with ui-config-service and ui-app-hub to show the primary and secondary navigation for your microapps
* `pxh-badge`, a small component for displaying "unread counts" for an item, with configurable color options (required by pxh-navigation)
* `pxh-login`, a fixed bottom-left component that offers login controls, user profile information, and access to user settings
* `pxh-overlay`, a semi-transparent background that overlays microapp content when the user opens the navigation drawer in mobile and tablet contexts
* `pxh-wrapper` - an element that wraps around the view header and view in your Microapp
* `pxh-view-header` - a fixed top header that should display a user's context within a microapp, potentially with a px-context-browser component
* `pxh-view` - a flexible, responsive target container for all microapp content
* `pxh-toasts` - Configurable messages to the user that "toast" in the upper-right corner of their web browser
* `pxh-notifications` - List of user's unacknowledged messages that persist in a menu accessible in the lower-left corner of their web browser

### [Screencast Introduction to pxh-chrome](https://ge.box.com/s/3ffmgwpntrbd6gjenw4zzqzkh1v2f387)
### [Introduction Slides](https://github.build.ge.com/pages/212326609/pxh-chrome-intro)

## Using pxh-chrome
### Use pxh-chrome on its own
#### Pre-work
1. Make sure your GE proxies are configured
1. Make sure you have Node installed
1. Install Bower and Gulp globally

#### Build pxh-chrome
1. Download, clone or fork pxh-chrome to your local machine
1. Run `npm install && bower install` at the command line to install all dependencies
1. Run `gulp` to compile the static files for pxh-chrome
1. Run `gulp serve` to start your local development server
  * Type `CTRL+C` to shut down your local development server
1. Navigate to [http://localhost:4000](http://localhost:4000) to view a static version of pxh-chrome, which shows the styles, markup and responsive behavior of pxh-chrome
  * The navigation links are for demonstration purposes, and will not allow you to navigate between pages

**Need a more robust prototyping environment?** Try [Propeller-Px](https://github.build.ge.com/212326609/propeller-px), which includes pxh-chrome along with server-side templating, jQuery, Bootstrap, GitHub Pages hosting, etc.

### Use pxh-chrome in your project
#### Use the CDN
The easiest way to drop pxh-chrome into your project is via the CDN. Simply drop `pxh-chrome.css` and `pxh-chrome.js` into your HTML file with the following lines:

```html
<link rel="stylesheet" href="//dzlpbrbc7yvq0.cloudfront.net/pxh-chrome/latest/dist/css/pxh-chrome.min.css">
```

```html
<script src="//dzlpbrbc7yvq0.cloudfront.net/pxh-chrome/latest/dist/js/pxh-chrome.min.js"></script>
```

Then, update your HTML to use pxh-chrome's markup and classes. Refer to the source code of `index.html` for examples.

#### Use Bower
Add pxh-chrome to your project using Bower:

```bash
bower install https://github.build.ge.com/hubs/pxh-chrome --save
```

Add `pxh-chrome.css` and `pxh-chrome.js` to your HTML file:

```html
<link rel="stylesheet" href="bower_components/pxh-chrome/dist/css/pxh-chrome.css">
```

```html
<script src="bower_components/pxh-chrome/dist/js/pxh-chrome.js"></script>
```

Then, update your HTML to use pxh-chrome's markup and classes. Refer to the source code of `index.html` for examples.

#### What files should I use?
The following CSS files are available on the CDN or as a local dependency via Bower:

* `pxh-chrome.min.css` - Default pxh-chrome styles in minified, compressed CSS (recommended in production)
* `pxh-chrome.css` - Default styles in expanded, readable, commented CSS (recommended for development)
* `pxh-chrome-rtl.min.css` - Experimental pxh-chrome layout for right-to-left languages (minified)
* `pxh-chrome-rtl.css` - Experimental layout for right-to-left-languages (expanded)

The following JS files are available via CDN or Bower:

* `pxh-chrome.min.js` - pxh-chrome scripts in minified, compressed Javascript (recommended in production)
* `pxh-chrome.js` - Scripts in expanded Javascript (recommended for development)

## Design Philosophy
Harry Roberts' [discovr](https://github.com/csswizardry/discovr/tree/correct) application from his CSS Architecture workshop is a great reference that encapsulates much of our markup and style goals with pxh-chrome.

pxh-chrome aims to provide the AppHub with enough HTML, CSS and JS to render the above components, and then get out of the way for microapp development. In achieving this goal it loads a few external depenencies (detailed in its `bower.json` file), including Sass components from Px, Font Awesome, and the GE Inspira Sans typeface.

We use Sass to author our style rules for pxh-chrome, which in turn are compiled into CSS. We pull in our color variables from [px-colors-design](https://github.com/PredixDev/px-colors-design), reset and normalize the browser's default styles with [normalize.css](https://github.com/PredixDev/px-normalize-design), and [update the global `box-sizing` state](https://github.com/PredixDev/px-box-sizing-design) to `border-box` for all elements. Additionally we use [sass-mq](https://github.com/sass-mq/sass-mq) to define our responsive media queries, ensuring they are consistent and readable.

We import [px-typography](https://github.com/PredixDev/px-typography-design) and [px-icongraphy](https://github.com/PredixDev/px-iconography-design) to bring in the GE Inspira fonts and Font Awesome icons. Since these dependencies are available in the global space you can likely eliminate them from your Microapp, reducing your page weight without any ill effect.

### Mobile-First
pxh-chrome adheres to a "mobile first" philosophy, where we consider the mobile experience to be the bedrock of our application. The "mobile" view is the "default" view of our application, and so all our styles start from these base rules. As the width of our application increases, we add additional styles as necessary to optimize our design and layout for wider and wider displays, including tablets, laptops and desktops.

When coding in a "mobile first" environment you must always ask yourself the question: "What rules do I need to add to [narrower responsive breakpoint] in order to get it to look right at [wider responsive breakpoint]?" The more consistently you layer styles _on top of_ narrower breakpoints, the easier your code will be to read, understand and reason with.

### Media Queries
We use [sass-mq](https://github.com/sass-mq/sass-mq). `@md`, `@lg`, etc.

Define your responsive styles using `$from` (mobile-first philosophy) rather than `$until` (desktop-first philosophy) ... Work hard to maintain this consistency; be prepared to rethink and refactor your code to make this happen.

### BEM
pxh-chrome uses [BEM syntax when naming CSS classes](http://csswizardry.com/2013/01/mindbemding-getting-your-head-round-bem-syntax/). BEM (Block, Element, Modifier) gives us a controlled grammer for describing the relationships between classes, which is incredibly helpful when working with large, sophisticated design systems like Px.

* `block` -- person
* `block--modifier` -- person--female, person--jogging
* `block--modifier@md` -- menu--hidden@md
* `block__element` -- person__arm, person__hand, person__finger
* `block__element--modifier` -- person__arm--raised, person__finger--pointing
* `block--modifier-until@lg` -- menu--collapsed-until@lg
* `block--modifier@md-until@lg` -- menu--fixed@md-until@lg
* `block__element--modifier@md` -- menu__summary--visible@md
* `block__element--modifier@md-until@lg` -- menu__summary--visible@md-until@lg

### Inverted Triangle (ITCSS)
We follow the Inverted Triangle architecture for our Sass manifest. Start with the most general selector rules at the top of your manifest, and get more and more specific as you move towards the bottom.

The triangle is divided into layers. Settings, Tools, Generic, Base, Objects, Components, "Trumps" (let's call them Overrides or Utilities instead).

Make sure you put your styles in the right layer. Most of pxh-chrome's styles are in the Components layer.

See Harry Roberts' slides from [Managing CSS Projects with ITCSS](https://speakerdeck.com/dafed/managing-css-projects-with-itcss) for more information.

### Classes are cheap. Your time is not.
HTML and classes are cheap. Your time is not cheap. The time of developers trying to understand your code in the future is not cheap. Write classes that are verbose (enough) and are self-documenting.

Classes should be readable. We use a lot of classes, especially with BEM. Especially with responsiveness.

Don't nest classes. Don't rely on inheritance. Write your styles and classes to be as shallow as possible.

Refer to [CSS Guidelines](http://cssguidelin.es/) for more best practices, including [recommendations against using nested selectors](http://cssguidelin.es/#nesting).

### Principle of Immutability
* A class should look and behave the same way no matter where it's used in your markup and no matter what breakpoint you're at.
* Want an element to look different in a different context? Write a "modifier" class and add that class to it
* Want an element to look different at a particular breakpoint? Write a "responsive modifier" class and add that class to it

### Theming, Branding and Personalization
There are a number of ways to skin this cat, from redefining variables at the pxh-chrome level and recompiling your CSS to embedding your theme styles in the <head> of your HTML file. Refer to Harry Roberts' [4 1/2 Methods for Theming in (S)CSS](https://speakerdeck.com/csswizardry/4half-methods-for-theming-in-s-css).

### Triggering "Chromeless" Mode
To trigger "chromeless" mode, which hides the drawer, drawer toggle, navigation, and login module, simply:

1. Replace the `pxh-chromeful` class on your `<body>` element with `pxh-chromeless`
1. Remove the markup for `pxh-drawer`, `pxh-view-header-toggle` and `pxh-overlay` from your HTML file (we're not showing it so we don't need it)
1. Remove the `<script>` tag for `pxh-chrome.js`
  * `pxh-chrome.js` is not required for chromeless view, as the JavaScript is only necessary for managing drawer state between breakpoints, page refreshes and toggle events... if there's no drawer, there's no need for JavaScript!

## Working with toasts and notifications
Toasts are little messages for the user that appear at the bottom of the viewport on mobile, and at the upper-right corner of the browser window on tablet and mobile.

### Adding a new toast
Here's the most basic example:

```
toast.add({ key : 'value'});
```

Here it is in some event logic, which will create a new toast when the page loads:

```
document.addEventListener('DOMContentLoaded', function(event) {
  toast.add({ key : 'value'});
});
```

## Listening for resize events
Since pxh-chrome is mobile-first and responsive, there may be cases where you need to rerender your microapp content based on changes in width to the `pxh-view` wrapping element. This is common with SVGs or other types of content that need to render at a specific and known pixel width.

Listening to `window.width` will cover many responsive use cases, but it doesn't cover the expansion (and collapse) of the navigation drawer, as that behavior changes the width of `pxh-view` without changing the width of the browser window.

Fortunately, pxh-chrome and the AppHub provides a custom event, `pxhViewResized`, that fires whenever the `pxh-view` element changes size. Simply listen for this event in your microapp, and respond as necessary after it fires.

* We include the ResizeSensor class from css-element-queries to give ourselves the ability to track when elements resize
* We grab `pxh-view` by ID and put it in a var
* `document.getElementById('js-view'); // you'll need to add this ID to your markup`
* In cases where pxh-chrome uses classes or IDs to grab elements via JS, we prefix them with `js-*` so we know there are no actual CSS styles associated with that attribute
* We create a new custom event called `pxhViewResized`

```javascript
var pxhViewResized = new CustomEvent('pxhViewResized', {
 'detail' : pxhView.offsetWidth
});
```

* Since it's a custom event, we return the offsetWidth in the detail object
* We add a new ResizeSensor to the pxh-view element, which dispatches the pxhViewResized event when it fires

```javascript
new ResizeSensor(pxhView, function() {
  document.dispatchEvent(pxhViewResized);
});
```

**Here's a basic usage example (do this in your microapp ... pxh-chrome should handle the rest):**

```javascript
document.addEventListener('pxhViewResized', function(event) {
  console.log('pxhViewResized was fired!');
  console.log('width changed to ' + event.detail);
});
```

P.S. If you look at the console output for this example, you'll realize you'll want to implement some debouncing on your event listener ... our current thinking is that the debouncing should be the responsibility of the microapp, not the event itself, but we're open to alternative perspectives.

## Running Tests
**To run unit tests** type the following at the command line:

```bash
gulp serve:test
```

This will start a local testing server at [http://localhost:4040](http://localhost:4040) and automatically open a browser window that shows the results of all the unit tests located at `./test/unit/spec/**/*.js`.

Type `CTRL+C` to shut down the unit testing server.

**To run end-to-end tests,** type the following at the command line:

```bash
gulp serve:e2e
```

This will install (or update, if already installed) a local version of Selenium WebDriver for Protractor, and then start a local development server at [http://localhost:4444](http://localhost:4444).

As soon as the server finishes loading (watch your Terminal output) open a new Terminal window and type the following:

```bash
gulp e2e
```

This will run all the end-to-end tests located at `./test/e2e/spec/**/*.js` and show you the results.

When you're done, type `CTRL+C` to shut down the development server.

**Note:** If you need to manually install WebDriver you can do so with this command:

```bash
./node_modules/protractor/bin/webdriver-manager update
```

## Browser Support
pxh-chrome is tested in the most recent major releases of Chrome, Firefox, Safari, iOS Safari, and Internet Explorer (IE11 on Windows 7 and IE Edge on Windows 10).

## Distributing pxh-chrome
1. Run `gulp` to compile your local version of pxh-chrome to your `/dist` folder
  * The files at `/dist` contain the compiled, concatenated and minified HTML, CSS and JavaScript for pxh-chrome
1. Run `gulp serve:dist` to test your distribution version of pxh-chrome for any last-minute bugs
  1. At this point you might want to test your local version of pxh-chrome in your local version of ui-app-hub
  1. In your local pxh-chrome folder, type `bower link` to create a local symlink for pxh-chrome
  1. In your local ui-app-hub folder, type `bower link pxh-chrome` to tell ui-app-hub to use _your local modified version_ of pxh-chrome rather than the version listed on ui-app-hub's `bower.json` file
  1. In your local ui-app-hub folder, run `gulp` (or `node app.js` or `nodemon`) to start your local AppHub
  1. Preview your changes to your local ui-app-hub and local pxh-chrome at [http://localhost:4000](http://localhost:4000)
  1. If anything looks broken, keep modifying your local ui-app-hub and pxh-chrome until it looks right
1. If everything looks good, bump your local version of pxh-chrome (use the `gulp bump` task with parameters listed above), merge to master, tag it as a release, and push it to GitHub (remember to push your `--tags` too!)
1. Update the `bower.json` in your local version of ui-app-hub to reference your latest version of pxh-chrome
1. Run `bower install` to install the version of pxh-chrome that you just pushed to GitHub
1. Fire up your local AppHub
1. If everything looks good, commit your changes to ui-app-hub and push them to GitHub

## Creating a new release of pxh-chrome
The release process has a few steps, [which we've carefully documented here](https://github.build.ge.com/gist/212326609/7299a13d2054479b2f3c).

Each time you cut a new release of pxh-chrome you're going to:

* Write really good release notes in the changelog
* Finalize and commit your source files
* Generate a new `/dist`
* Run unit and end-to-end tests
* Commit your dist files
* **Finish your release and push it to GitHub**
* Publish your (really good) release notes on GitHub
* **Push your release to the CDN**


CHANGELOG.md
# pxh-chrome changelog

### 3.0.4 - Feb, 2018
* Local fonts

### 3.0.3 - March 7, 2017
* Update colors to latest px 2017 themes.

### 3.0.1 - December 15, 2016
* Remove `.tmp` folder
  * Everything now builds to and serves from the `/dist` folder
* Update Sass compilation gulp tasks to compile against the flattened Sass files in `/dist/sass`
  * This way everything is compiling against the exact same Sass files, no matter if it's being done within pxh-chrome, or by ui-theme-service
* Remove `includePaths` property from Sass compilation tasks
  * We don't need them now that we're compiling against flattened Sass
* Remove underline on hover by default for all links
* Bump browser-sync, eslint, gulp-sass, gulp-sourcemaps

### 3.0.0 - December 9, 2016
**In order to support future theming-as-a-service capabilities, this release introduces breaking changes that will affect apps that use the pxh-chrome Sass files directly.**

* The file `sass/_settings.theme-default.scss` has been renamed `sass/_theme.default.scss`
* The file `sass/_settings.theme-prechrome.scss` has been renamed `sass/_theme.prechrome.scss`

Apps that use the compiled pxh-chrome CSS are not affected.

* Add import-resolve and `flatten.js` task to generate concatenated Sass manifest files
* Hard code `../bower_components` path to bower components to facilitate manifest concatenation
* Add `theme` layer to all Sass manifests and move theme-related Sass partials into it
* Refine prechrome theme so it follows the theming contract
* Add text decoration and hover style properties for default hyperlinks, and expose them as variables to allow customization via theming
* Add `sass:flatten` Gulp task to flatten Sass manifests and output them to `dist/sass
* Tweak view header font size modifier so it renders at 20px by default
* Bump versions on babel-core, eslint, protractor

### 2.11.1 - November 30, 2016
* Add unit tests that test the behavior of toast and notification default strings with (and without) corresponding `window.nav.localeData` values
* Tweak `.eslintc` overrides for eslint-config-airbnb-base
* Resolve all ESLint errors in `pxh-chrome.js`

### 2.11.0 - November 29, 2016
* Add support for localization/internationalization of hard-coded default toast strings via App Hub's `window.nav.localeData` object
  * If `text` or `actionLabel` are defined in the passed toast object, it simply uses those values
  * If not, it looks for `window.nav.localeData.toastTextDefault` for the default internationalized "You received a new notification." message and `window.nav.localeData.toastActionLabelDefault` for the default internationalized "Action" label
  * If not, it uses the default strings hard-coded in pxh-chrome
* Fix height of view header drawer toggle to match drawer toggle
* Bump versions on browser-sync, eslint, gulp-sourcemaps, mocha

### 2.10.0 - November 18, 2016
* Replace "show more/less" text buttons in toasts and notifications with "down/up chevrons" to simplify internationalization
* Bump versions on browser-sync, eslint, handlebars, lodash, protractor
* Update `yarn.lock` file with new node modules
* Fix unmerged feature branch from 2.9.0 release

### 2.9.0 - November 18, 2016
* _(this was an incomplete release)_
* Add `latest` tag to CDN, which always points to the latest available version of pxh-chrome
  * Use it like this:
    * `<link rel="stylesheet" href="//dzlpbrbc7yvq0.cloudfront.net/pxh-chrome/latest/dist/css/pxh-chrome.min.css">`
    * `<script src="//dzlpbrbc7yvq0.cloudfront.net/pxh-chrome/latest/dist/js/pxh-chrome.min.js"></script>`
* Update README docs to use `latest` tag on CDN

### 2.8.2 - November 15, 2016
* Add color variables for new 2017 cooler shades of gray
  * Namespaced as `$l-grayxx` (`l` for `light` as opposed to `d` for `dark` theme)
* Add color variables for dark "lights out" Predix UI theme
* Add "alpha" version of `_settings.theme-lights-out.scss` "lights out" theme
  * It's not ready yet but we want to start testing it
* Remove a stray `$white` variable for drawer header link color so pxh-chrome Sass files are completely independent from colors defined in `_settings.colors.scss`
  * The only place these colors are used now are in theme files (e.g. `_settings.theme-defaults.scss`, `_settings.theme-lights-out.scss`)
* Use `$white` and `$black` variables instead of hard-coded RGB values for box shadows and borders
* Fix end-to-end test for pxh-chrome version number

### 2.8.1 - November 10, 2016
* Update demo text for "user with a really long name"
* Replace avatar image with Thomas Edison

### 2.8.0 - November 7, 2016
* Change default `$pxh-view-bg-color` value to `transparent` instead of a color to minimize confusion and surprise when applied to existing designs

### 2.7.0 - November 7, 2016
* Change default value of `$pxh-view-header-border-color` from an actual color to black with 30% alpha transparency, to support a wider variety of `pxh-view` background colors
* Add `$pxh-view-bg-color` variable to allow customization of default `pxh-view` background color independently from default `<html>`/`<body` color or default `pxh-wrapper` color
* Bump required version of node to 6.3.0 and add required version of npm
* Bump node modules
* Bump sass-mq to 3.3.1
* Add carets to bower component references so latest minor releases will always be installed (yolo)
* Fix path to `sass` folder in Sass watch task

### 2.6.0 - October 26, 2016
* Update Font Awesome to [4.7.0](https://github.com/FortAwesome/Font-Awesome/pull/10012)
  * New icons including handshake, open envelope, bath, shower, id card, address book, thermometer
* Bump Babel node modules
* Tweak 2.5.0 release notes

### 2.5.0 - October 24, 2016
* Move prechrome override variables out of the Sass manifest and into a `settings.theme-prechrome` layer where they belong
* Split Px colors and font/line/spacing Sass values into their own `settings` partials

test/e2e/protractor.config.js
// protractor.config.js
/*
* Defines protractor configuration for end-to-end tests
* */

exports.config = {
  framework: 'mocha',
  directConnect: true,
  specs: [],
  suites: {
    smoke:      'spec/smoke/*.spec.js',
    nav:        'spec/drawer/*.spec.js',
    responsive: 'spec/responsive/*.spec.js',
    toast:      'spec/toast/*.spec.js',
  },
  capabilities: {
    'browserName': 'chrome' // chrome or firefox ... directConnect doesn't support safari
  },
  mochaOpts: {
    ui: 'bdd',
    reporter: 'list',
    timeout: '10000'
  }
}


test/e2e/spec/common/common.po.js
/*
 * commonPage
 * This page object contains commonly used utilities
 *
 * */

'use strict';

(function () {

  var commonPage = function () {

    return {
      //check to see if an element has such css class provided
       hasClass: function (ele, cls) {
        return ele.getAttribute('class').then(function (classes) {
          return classes.split(' ').indexOf(cls) !== -1;
        });
      },

      convertPxToNum: function (px) {
        return parseInt(px, 10);
      },

      parseColor: function (input) {
        if (input === 'transparent') {
          return input;
        }
        else if (input) {
          var regex = /^rgba\s*\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*,\s*([\.\d]+)\s*\)$/i
          var matches = regex.exec(input);
          if (matches) return matches;
        }
        else throw new Error("Color " + input + " could not be parsed.");
      },

      isOpaque: function (input) {
        if (input === 'transparent') {
          return false;
        }
        else if (input) {
          var regex = /^rgba\s*\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*,\s*([\.\d]+)\s*\)$/i
          var matches = regex.exec(input);
          if (parseFloat(matches[4]) === 0) return false;
          else {
            return true;
          }
        }
        else throw new Error("Color " + input + " could not be parsed.");
      },

      // To generate a random String name
      getRandomString: function () {
        var chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXTZabcdefghiklmnopqrstuvwxyz";
        var string_length = 8;
        var randomstring = '';
        for (var i = 0; i < string_length; i++) {
          var rnum = Math.floor(Math.random() * chars.length);
          randomstring += chars.substring(rnum, rnum + 1);
        };
        return randomstring;
      },

      waitForElement: function (elementToWait) {
        var EC = protractor.ExpectedConditions;
        var microappContent = element(by.css(elementToWait));
        browser.wait(EC.presenceOf(microappContent), 10000);
      },

      waitForUrlToChange: function (urlToWait) {
        var currentUrl;
        return browser.getCurrentUrl()
        .then(function storeCurrentUrl(url) {
          currentUrl = url;
          })
        .then(function waitForUrlToChangeTo() {
           return browser.wait(function waitForUrlToChangeTo() {
           return browser.getCurrentUrl().then(function compareCurrentUrl(url) {
             return urlToWait == url;
           });
           });
         });
      },

      navigateToUrl: function (url) {
        browser.driver.get(url);
        this.waitForUrlToChange(url);
      },

      getErrorMessage: function () {
        //
      },

      getSuccessMessage: function () {
        //
      },

      // Choose SelectField
      // Click an option under a select tag
      // Passing a selector, index of the option, and text that matches the text under the select menu
      selectSelectorField : function (selector, index , text){
        return element(selector).all(by.css('option')).then(function (eles) {
          return eles[index].getText().then(function doesOptionMatch(txt){
            if (txt === text){
              return eles[index].click();
            }
          });
        });
      }
    };
  };
  module.exports = new commonPage();
}());


build.sh
BUILD=$1
PREFIX=$(node -p "require('./package.json').version")
VERSION=$PREFIX.$BUILD

export http_proxy=http://sjc1intproxy01.crd.ge.com:8080
export https_proxy=$http_proxy
export all_proxy=$http_proxy
export no_proxy="localhost, 127.0.0.1, *.ge.com"

# Set uppercase versions of these variables (needed by cdnify.js at time of writing).
export HTTP_PROXY=$http_proxy
export HTTPS_PROXY=$http_proxy
export ALL_PROXY=$http_proxy
export NO_PROXY="$no_proxy"

echo ---------------------------
echo Environment Variables START
echo ---------------------------
env
echo -------------------------
echo Environment Variables END
echo -------------------------

export NVM_DIR=~/.nvm
source ~/.nvm/nvm.sh

echo ----------------
echo Versions of nvm:
echo ----------------
nvm --version

echo
echo -------------------------------
echo Use latest 6.x version of node:
echo -------------------------------
nvm use v6.9.1

echo
rm -rf /root/.npm/*.lock.STALE
rm -rf node_modules
echo
echo -------------
echo Node version:
node -v
echo -------------
echo ''
echo ------------
echo npm version:
npm -v
echo ------------

echo -----------------
echo NPM INSTALL START
echo -----------------
npm install
echo -----------------
echo NPM INSTALL END
echo -----------------

echo ----------------------
echo BOWER cach clean START
echo ----------------------
bower cache clean
echo --------------------
echo BOWER cach clean END
echo --------------------

echo -------------------
echo BOWER INSTALL START
echo -------------------
bower install --force-latest
echo -----------------
echo BOWER INSTALL END
echo -----------------


# Copy static assets to CDN and change template strings
# echo -----------------
# echo Running cdnify.js
# echo -----------------
# node cdnify.js -v $VERSION

# Generate /dist for Jenkins
echo ------------------
echo RUNNING GULP
echo ------------------
gulp

cd dist
# Create zip file for artifactory (and Jenkins)
zip -r ../bxjoc-pxh-chrome-$VERSION.zip *
cd ..
ls -lrtR


gulpfile.babel.js
// generated on 2016-03-14 using generator-webapp 2.0.0
import gulp from 'gulp';
import gulpLoadPlugins from 'gulp-load-plugins';
import browserSync from 'browser-sync';
import del from 'del';
import path from 'path';
import replace from 'gulp-replace';
import ext_replace from 'gulp-ext-replace';
import shell from 'gulp-shell';
import extReplace from 'gulp-ext-replace';
import gulpsmith from 'gulpsmith';
import layouts from 'metalsmith-layouts';
import copy from 'metalsmith-copy';
import handlebars from 'handlebars';
import lodash from 'lodash';

const $ = gulpLoadPlugins();
const reload = browserSync.reload;

const componentConfig = {
  site: {
    title: 'pxh-chrome',
    version: '3.0.1',
  },
};

gulp.task('sass', ['sass:flatten'], () => {
  gulp.start('sass:min');
  return gulp.src('dist/sass/*.scss')
    .pipe($.plumber())
    .pipe($.sourcemaps.init())
    .pipe($.sass.sync({
      outputStyle: 'expanded',
      precision: 10,
    }).on('error', $.sass.logError))
    .pipe($.autoprefixer({ browsers: ['> 1%', 'last 2 versions', 'Firefox ESR'] }))
    .pipe($.sourcemaps.write('.'))
    .pipe(gulp.dest('dist/css'))
});

gulp.task('sass:min', () => {
  return gulp.src('dist/sass/*.scss')
    .pipe($.plumber())
    .pipe($.sass.sync({
      outputStyle: 'expanded',
      precision: 10,
    }).on('error', $.sass.logError))
    .pipe($.autoprefixer({ browsers: ['> 1%', 'last 2 versions', 'Firefox ESR'] }))
    .pipe($.cssnano())
    .pipe(extReplace('.min.css', '.css'))
    .pipe(gulp.dest('dist/css'));
});

gulp.task('sass:flatten', shell.task('node ./flatten.js'));

gulp.task('js', () => {
  return gulp.src('public/js/**/*.js')
    .pipe($.plumber())
    .pipe($.sourcemaps.init())
    .pipe($.babel())
    .pipe($.sourcemaps.write('.'))
    .pipe(gulp.dest('dist/js'))
    .pipe($.if('*.js', $.uglify({
      preserveComments: 'some',
    })))
    .pipe(extReplace('.min.js', '.js'))
    .pipe(gulp.dest('dist/js'))
    .pipe(reload({ stream: true }));
});

function lint(files, options) {
  return () => {
    return gulp.src(files)
      .pipe(reload({ stream: true, once: true }))
      .pipe($.eslint(options))
      .pipe($.eslint.format());
  };
}
const testLintOptions = {
  env: {
    mocha: true,
  },
};

gulp.task('lint', lint('public/js/**/*.js'));
gulp.task('lint:test', lint('test/unit/spec/**/*.js', testLintOptions));

gulp.task('smith', () => {
  gulp.src(['src/screens/*'])
  .pipe($.frontMatter()).on('data', (file) => {
    lodash.assign(file, file.frontMatter);
    delete file.frontMatter;
  })
  .pipe(
    gulpsmith()
    .metadata(componentConfig)
    .on('error', console.log.bind(console))
    .use(layouts({
      engine: 'handlebars',
      directory: 'src/layouts',
      pattern: '*.hbs',
      default: 'default.hbs',
      partials: 'src/partials',
    }))
    .on('error', console.log.bind(console))
    .use(copy({
      pattern: '*.hbs',
      extension: '.html',
      move: true,
    }))
    .on('error', console.log.bind(console))
  )
  .pipe(gulp.dest('dist'))
  .pipe(reload({ stream: true }));
});

gulp.task('extras', ['fonts'], () => {
  return gulp.src([
    'public/*.*',
    '!public/*.html',
    '!public/.eslintrc',
  ], {
    dot: true,
  }).pipe(gulp.dest('dist'));
});

gulp.task('img', () => {
  return gulp.src(['public/img/*']).pipe(gulp.dest('dist/img'));
});
gulp.task('fonts', () => {
  return gulp.src(['fonts/*']).pipe(gulp.dest('dist/fonts'));
});

gulp.task('clean', del.bind(null, ['dist']));

gulp.task('serve', ['sass', 'js', 'extras', 'img'], () => {
  browserSync.init({
    ui: {
      port: 4040,
    },
    port: 4000,
    notify: false,
    server: {
      baseDir: ['dist', 'public'],
      routes: {
        '/bower_components': 'bower_components',
      },
    },
  });

  gulp.watch('src/**/*.hbs', ['smith']);
  gulp.watch('sass/**/*.scss', ['sass']);
  gulp.watch('public/js/**/*.js', ['js']);
  gulp.watch('public/img/**', ['img']);

  gulp.watch([
    'dist/*.html',
    'dist/img/*',
    'dist/css/*.css',
    'dist/js/*.js',
  ]).on('change', reload);
});

gulp.task('webdriver:update', shell.task('./node_modules/protractor/bin/webdriver-manager update'));

gulp.task('e2e', shell.task('protractor ./test/e2e/protractor.config.js'));

gulp.task('serve:e2e', ['webdriver:update', 'sass', 'js', 'extras', 'img'], () => {
  browserSync.init({
    ui: false,
    port: 4444,
    open: false,
    notify: false,
    server: {
      baseDir: ['dist'],
      routes: {
        '/bower_components': 'bower_components',
      },
    },
  });
});

gulp.task('serve:test', ['js'], () => {
  browserSync.init({
    ui: false,
    port: 4040,
    notify: false,
    server: {
      baseDir: 'test/unit',
      routes: {
        '/js': 'dist/js',
        '/bower_components': 'bower_components',
      },
    },
  });

  gulp.watch('public/js/**/*.js', ['js']);
  gulp.watch('test/unit/spec/**/*.js').on('change', reload);
  gulp.watch('test/unit/spec/**/*.js', ['lint:test']);
});

gulp.task('build', ['lint', 'smith', 'sass', 'js', 'extras', 'img']);

gulp.task('dist', ['clean'], () => {
  gulp.start('build');
});

// the default task is the dist task
gulp.task('default', ['dist']);
